src/Assimp/Silk.NET.Assimp/CustomFileIO.cs

```diff
@@ -0,0 +1,198 @@
+        *file = new File
+        {
+            ReadProc = (delegate* unmanaged[Cdecl]<File*, byte*, nuint, nuint, nuint>)&ReadFile,
+            WriteProc = (delegate* unmanaged[Cdecl]<File*, byte*, nuint, nuint, nuint>)&WriteFile,
+            FileSizeProc = (delegate* unmanaged[Cdecl]<File*, nuint>)&FileSize,
+            TellProc = (delegate* unmanaged[Cdecl]<File*, nuint>)&FileTell,
+            FlushProc = (delegate* unmanaged[Cdecl]<File*, void>)&FileFlush,
+            SeekProc = (delegate* unmanaged[Cdecl]<File*, nuint, Origin, Return>)&FileSeek,
+            UserData = (byte*)GCHandle.ToIntPtr(gch)
+        };
+        
+        return file;
+    }
+    
+    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
+    private static void CloseFile(FileIO* mFileSystem, File* pFile)
+    {
+        var sHandle = GCHandle.FromIntPtr((nint) pFile->UserData);
+        if (sHandle.Target is not Stream s)
+        {
+            throw new InvalidOperationException("Invalid UserData for FileIO.");
+        }
+        
+        s.Dispose();
+        sHandle.Free();
+        SilkMarshal.Free((nint) pFile);
+    }
+    
+    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
+    private static nuint ReadFile(File* file, byte* buffer, nuint size, nuint count) =>
+        GCHandle.FromIntPtr((nint) file->UserData).Target is not Stream s
+            ? throw new InvalidOperationException("Invalid UserData for File.")
+            : (nuint) s.Read(new Span<byte>(buffer, (int) (size * count))) / size;
+    
+    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
+    private static nuint WriteFile(File* file, byte* buffer, nuint size, nuint count)
+    {
+        if (GCHandle.FromIntPtr((nint) file->UserData).Target is not Stream s)
+        {
+            throw new InvalidOperationException("Invalid UserData for File.");
+        }
+
+        s.Write(new Span<byte>(buffer, (int) (size * count)));
+        return count;
+    }
+    
+    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
+    private static nuint FileSize(File* file) =>
+        GCHandle.FromIntPtr((nint) file->UserData).Target is not Stream s
+            ? throw new InvalidOperationException("Invalid UserData for File.")
+            : (nuint) s.Length;
+    
+    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
+    private static nuint FileTell(File* file) =>
+        GCHandle.FromIntPtr((nint) file->UserData).Target is not Stream s
+            ? throw new InvalidOperationException("Invalid UserData for File.")
+            : (nuint) s.Position;
+    
+    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
+    private static void FileFlush(File* file)
+    {
+        if (GCHandle.FromIntPtr((nint) file->UserData).Target is not Stream s)
+        {
+            throw new InvalidOperationException("Invalid UserData for File.");
+        }
```

Comment:
This function can probably be wrapped into a `Stream s = GetStreamFromUserData(file->UserData);`, which would de-duplicate a lot of code here